airflow/timetables/events.py

```diff
@@ -0,0 +1,111 @@
+        the current time
+    :param description: A name for the timetable to display in the UI. Default None will be shown as
+                        "X Events" where X is the len of event_dates
+    """
+
+    def __init__(
+        self,
+        event_dates: Iterable[DateTime],
+        restrict_to_events: bool = False,
+        presorted: bool = False,
+        description: Optional[str] = None,
+    ):
+
+        self.event_dates = list(event_dates)  # Must be reversible and indexable
+        if not presorted:
+            # For long lists this could take a while, so only want to do it once
+            self.event_dates = sorted(self.event_dates)
+        self.restrict_to_events = restrict_to_events
+        if description is None:
+            self.description = (
+                f"{len(self.event_dates)} Events between {self.event_dates[0]} and {self.event_dates[-1]}"
+            )
+            self._summary = f"{len(self.event_dates)} Events"
+        else:
+            self._summary = description
+            self.description = description
+
+    @property
+    def summary(self) -> str:
+        return self._summary
+
+    def __repr__(self):
+        return self.summary
+
+    def next_dagrun_info(
+        self,
+        *,
+        last_automated_data_interval: Optional[DataInterval],
+        restriction: TimeRestriction,
+    ) -> Optional[DagRunInfo]:
+        if last_automated_data_interval is None:
+            next_event = self.event_dates[0]
+        else:
+            future_dates = itertools.dropwhile(
+                lambda when: when <= last_automated_data_interval.end, self.event_dates  # type: ignore
+            )
+            next_event = next(future_dates, None)  # type: ignore
+            if next_event is None:
+                return None
+
+        return DagRunInfo.exact(next_event)
+
+    def infer_manual_data_interval(self, *, run_after: DateTime) -> DataInterval:
+        # If Timetable not restricted to events, run for the time specified
+        if not self.restrict_to_events:
+            return DataInterval.exact(run_after)
+
+        # If restricted to events, run for the most recent past event
+        # or for the first event if all events are in the future
+        if run_after < self.event_dates[0]:
+            return DataInterval.exact(self.event_dates[0])
+        else:
+            past_events = itertools.dropwhile(lambda when: when > run_after, self.event_dates[::-1])
+            most_recent_event = next(past_events)
+            return DataInterval.exact(most_recent_event)
+
+    def serialize(self):
+        return {
+            "event_dates": [str(x) for x in self.event_dates],
+            "restrict_to_events": self.restrict_to_events,
+        }
+
+    @classmethod
+    def deserialize(cls, data) -> Timetable:
+        return cls(
+            [pendulum.parse(x) for x in data["event_dates"]],  # type: ignore
```

Comment:
Why is this ignore needed? 🤔

Reply:
Otherwise mypy thinks there could be a None in there.

Reply:
In where? `data` is Any so the None check shouldn’t kick in. If it is `pendulum.parse` we should check against that in case of corrupted serialisation.

Reply:
I don't understand what that check would look like. Should I remove the ignore locally and paste what mypy is unhappy about here?

Reply:
Sorry, I got mixed up. Here's the error:

```
airflow/timetables/events.py:108: error: List comprehension has incompatible
type List[Union[Date, Time, DateTime, Duration]]; expected List[DateTime]
[misc]
                [pendulum.parse(x) for x in data["event_dates"]],
```

So it thinks that parse could return something other than a datetime, but since all we're serializing is datetimes, I don't think that will happen. Is there a better way to handle this?

Reply:
Maybe use `pendulum.DateTime.fromisotimestamp()` instead?